docs(installer): C4 / sequence / data-view HLD artifact set (w1qls.9) (#117)
scotthamilton77/agents-config: docs/architecture/installer/data-view.md

    StagingPlan {
        Tool Tool                "which tool this plan targets"
        Map  Items               "dict[Path, StagedItem] — IN-MEMORY, no temp-dir"
    }

    StagedItem {
        Path     Dest        "destination path, relative to the tool's dest_dir"
        bytes    Content     "post-flatten, post-transform bytes to write"
        FileKind Kind        "NAMESPACED_MD | SETTINGS_JSON | JSONC | TOML | OTHER | DIR"
        string   Namespace   "commands|skills|agents|rules|formulas | '' — 2nd merge key"
        bool     Executable  "chmod +x on write (e.g. beads scripts)"
    }

    Provenance {
        string Kind  "tool | plugin"
        string Name  "Tool enum value (tool) or plugin name string (plugin)"
    }

    IncludeDirective {
        string Kind  "file | all_rules — discriminated union (FileInclude | AllRulesInclude)"
        string Path  "FileInclude only — fragment path to inline; empty for all_rules"
    }

    Orphan {
        Tool     Tool       "which destination store"
        string   Namespace  "managed namespace or ''"
        Path     Path       "destination path on disk"
        FileKind Kind        "classification of the orphaned file"

>     }

Author: Fixed in 787718f — the `Orphan` entity now matches the dataclass: `tool: str` (the orphan bucket includes plugin namespaces like beads) and `kind: Literal["dir","file"]`, not `FileKind`.

    Counters {
        int Created     "new files written"
        int Written     "existing files overwritten"
        int Skipped     "hash-equal or user-declined"
        int BackedUp    "files copied before overwrite/prune"
        int Pruned      "orphans removed"
        int WouldCreate "dry-run mirror of Created — previewed, not written"
        int WouldUpdate "dry-run mirror of Written — previewed, not written"

>     }

Author: Fixed in 787718f — the `Counters` ER now lists the implemented fields `staged/created/updated/skipped/pruned/backed_up`; the invented `WouldCreate/WouldUpdate` are gone. The glossary line was aligned to the same names in 388154f.

- **`Provenance` carries the tool-vs-plugin asymmetry.** Tools are enum-keyed (`Tool` enum + adapter registry); plugins are string-keyed (dynamic discovery, no enum). `Provenance(kind, name)` lets a single `StagedItem` record either origin uniformly, so the merge engine can reason about "base asset vs plugin overlay" without caring which registry the item came from.
- **`IncludeDirective` is transient.** It is produced while `templates.py` flattens a `<!-- DYNAMIC-INCLUDE: … -->` marker (file form) or the ALL-RULES marker, then consumed immediately — the resulting flattened text lands in `StagedItem.Content`. It is modelled here because it is a named type in the data model (A.2), but it does not survive on the `StagedItem`. The `AllRulesInclude` variant carries no path: it expands the plan's already-staged rules collection, sorted and `\n---\n`-joined.
- **`FileKind` is an enum, not an entity.** Its six values are shown inline on `StagedItem.Kind`. It is the primary merge-dispatch key; `Namespace` is the secondary key (see the dispatch table).

## Merge-dispatch table — `(FileKind, namespace)` → `MergeStrategy`

> The collision matrix, keyed exactly as `core/merge/registry.py` keys it. `NAMESPACED_MD` is the only kind whose namespace changes the strategy; for every other kind the namespace component is unused and the lookup degenerates to a `FileKind`-only key.

Author: Fixed in 787718f — the dispatch table is now framed as the design contract that `core/merge/registry.py` *will* implement (Epic E; not yet built), not a description of current code, so it no longer cites a non-existent file as authoritative.

# Python Installer — Data View

> **Up**: [index](index.md)
> **Previous (reading order)**: [C4 L3 — Engine](c4-l3-engine.md)
> **Source bead**: `agents-config-w1qls.9`
> **Source spec**: [`installer-design.md`](installer-design.md) — §"Data model highlights", §"Configuration — installer.toml"

## Glossary

| Term | Meaning |
|---|---|
| `StagingPlan` | The aggregate root of the in-memory model: a `dict[Path, StagedItem]` (plus the target `Tool`). One is built per detected tool. **In-memory only** — it has no on-disk form in the operational path. |
| `StagedItem` | One planned destination file. The unit the merge + sync engines operate on. |
| `Provenance` | `(kind: "tool" | "plugin", name: str)` — preserves whether a `StagedItem` came from a tool's source tree or a plugin overlay, through the tool-vs-plugin registry asymmetry. |
| `FileKind` | The enum classifying a staged file. The **primary** merge-dispatch key. |

> | Namespace | The managed sub-dir (`commands` / `skills` / `agents` / `rules` / `formulas`) or `""`. The **secondary** merge-dispatch key. |

Author: Fixed in a005e2a — the `Namespace` glossary now says `None` (not `""`), matching the implemented `StagedItem.namespace: str | None`.

| `Orphan` | A prune candidate: on disk, absent from the plan, matching a retired glob. |
| `Counters` | The per-run tally (`staged` / `created` / `updated` / `skipped` / `pruned` / `backed_up`) surfaced in the exit summary. |
| Canonical ownership | Which actor is the source of truth for a piece of data: the human (source tree), the installer (plan + writes), or the tool (deployed store at runtime). |

## Purpose

Three complementary data views in one file:

1. **The in-memory model** (`Config`, `StagingPlan`, `StagedItem`, `Provenance`, `IncludeDirective`, `Orphan`, `Counters`) as an ER diagram — the shapes the engine builds and passes around. None of these persist; they live and die within one process invocation.
2. **The merge-dispatch table** — the `(FileKind, namespace)` → `MergeStrategy` lookup that the collision matrix is built on.
3. **The config + ownership boundaries** — the `installer.toml` schema, and which actor owns which data as it flows source → memory → disk → runtime.

The data view answers: *what shapes does the installer build in memory, what drives collision resolution, and who owns each piece of data along the way?*

## In-memory model ER diagram

> Field names below mirror the **implemented** dataclasses in `packages/installer/src/installer/core/model.py` verbatim (snake_case). Where a type is not yet built, the row is marked *design-forward* and traces to `installer-design.md`, not to current code.

Author: Fixed in a005e2a — the sentence now notes `Config` is the implemented dataclass in `packages/installer/src/installer/config.py` (the others mirror `core/model.py`).

        bool      dry_run        "design-forward: --dry-run preview, no writes"
        PruneMode prune          "design-forward: none | prune | prune_only"
        Path      dump_stage     "design-forward: --dump-stage target (debug)"
        StrList   retired_globs  "design-forward: installer.toml [prune].retired"
        Map       tool_overrides "design-forward: installer.toml [tools]"
    }

    StagingPlan {
        Map  items  "dict[Path, StagedItem] — IN-MEMORY; silently overwrites on dup dest_relpath (caller routes to merge)"
        Tool tool   "which tool this plan targets"
    }

    StagedItem {
        Path     source_path   "source file the bytes derive from"
        Path     dest_relpath  "destination path, relative to the tool's install root (dict key)"
        FileKind kind          "DIR | SETTINGS_JSON | JSONC | TOML | NAMESPACED_MD | OTHER"
        string   namespace     "managed namespace or null — 2nd merge-dispatch key"
        bytes    content       "post-flatten bytes; null when kind == DIR (derived at sync)"
        bool     executable    "sync-phase mode bit 0o755 vs 0o644 (e.g. beads scripts)"

>     }

Author: Fixed in a005e2a — the `StagedItem` ER block now includes the `provenance: Provenance` attribute, alongside the existing has-1 relationship line.

    end

    subgraph TOOLS["Tool-owned at runtime"]
        T1[AI assistants + bd CLI]
    end

    S1 -->|"staging reads"| M1
    S2 -->|"config reads"| M1
    M1 -->|"sync writes (hash-gated)"| D1
    D1 -->|"backup before overwrite/prune"| B1
    D1 -->|"runtime read (async)"| T1
```

### Ownership rules (worth memorising)

| Data | Owner | Lifetime | Notes |
|---|---|---|---|
| Source tree (`src/user/`, `src/plugins/`) | **Human** (via repo) | Permanent | Installer reads, NEVER writes. The "always edit source" guarantee. |
| `installer.toml` | **Human** | Permanent | Read-only to the install path. |
| `StagingPlan` / `StagedItem` | **Installer** | One invocation | In-memory; gone when the process exits. `--dump-stage` materialises a throwaway copy. |
| `Counters` / `Orphan` list | **Installer** | One invocation | Surfaced in the exit summary; not persisted. |
| Destination stores | **Installer** writes → **Tool** reads | Permanent on disk | Single writer at install time; consumed asynchronously at each tool's runtime. |
| Backups | **Installer** | Permanent on disk | Write-only recovery; never read back by the installer. |

### Explicit non-ownership

> - The installer does **not** own source content — it copies and flattens it, but the human authoring the repo is canonical. A `StagedItem.Content` is a *derived* artifact (post-flatten, post-transform), not a source of truth.

Author: Fixed in a005e2a — corrected to lowercase `StagedItem.content` to match the dataclass field name.

# Python Installer — Data View

> **Up**: [index](index.md)
> **Previous (reading order)**: [C4 L3 — Engine](c4-l3-engine.md)
> **Source bead**: `agents-config-w1qls.9`
> **Source spec**: [`installer-design.md`](installer-design.md) — §"Data model highlights", §"Configuration — installer.toml"

## Glossary

| Term | Meaning |
|---|---|
| `StagingPlan` | The aggregate root of the in-memory model: a `dict[Path, StagedItem]` (plus the target `Tool`). One is built per detected tool. **In-memory only** — it has no on-disk form in the operational path. |
| `StagedItem` | One planned destination file. The unit the merge + sync engines operate on. |

> | `Provenance` | `(kind: "tool" | "plugin", name: str)` — preserves whether a `StagedItem` came from a tool's source tree or a plugin overlay, through the tool-vs-plugin registry asymmetry. |

Author: ```suggestion
| `Provenance` | `(kind: "tool" \| "plugin", name: str)` — preserves whether a `StagedItem` came from a tool's source tree or a plugin overlay, through the tool-vs-plugin registry asymmetry. |
```

Author: Applied in commit d387bea — escaped the pipe character in the Provenance type signature.

| `FileKind` | The enum classifying a staged file. The **primary** merge-dispatch key. |
| Namespace | The managed sub-dir (`commands` / `skills` / `agents` / `rules` / `formulas`) or `None` when the tool root itself owns the file. The **secondary** merge-dispatch key. |

> | `IncludeDirective` | A discriminated union (`FileInclude` | `AllRulesInclude`) produced **transiently** while flattening DYNAMIC-INCLUDE markers; consumed during staging, not persisted on the `StagedItem`. |

Author: ```suggestion
| `IncludeDirective` | A discriminated union (`FileInclude` \| `AllRulesInclude`) produced **transiently** while flattening DYNAMIC-INCLUDE markers; consumed during staging, not persisted on the `StagedItem`. |
```

Author: Applied in commit d387bea — escaped the pipe character in the IncludeDirective type signature.

| `Orphan` | A prune candidate: on disk, absent from the plan, matching a retired glob. |
| `Counters` | The per-run tally (`staged` / `created` / `updated` / `skipped` / `pruned` / `backed_up`) surfaced in the exit summary. |
| Canonical ownership | Which actor is the source of truth for a piece of data: the human (source tree), the installer (plan + writes), or the tool (deployed store at runtime). |

## Purpose

Three complementary data views in one file:

1. **The in-memory model** (`Config`, `StagingPlan`, `StagedItem`, `Provenance`, `IncludeDirective`, `Orphan`, `Counters`) as an ER diagram — the shapes the engine builds and passes around. None of these persist; they live and die within one process invocation.
2. **The merge-dispatch table** — the `(FileKind, namespace)` → `MergeStrategy` lookup that the collision matrix is built on.
3. **The config + ownership boundaries** — the `installer.toml` schema, and which actor owns which data as it flows source → memory → disk → runtime.

The data view answers: *what shapes does the installer build in memory, what drives collision resolution, and who owns each piece of data along the way?*

## In-memory model ER diagram

> Field names below mirror the **implemented** dataclasses in `packages/installer/src/installer/core/model.py` verbatim (snake_case), plus the implemented `Config` dataclass in `packages/installer/src/installer/config.py`. This represents target state, not necessarily current code.

Author: Addressed in commit c0cd7f6 — preamble now reads 'Field names mirror the **target** dataclasses… This represents target state, not necessarily current code.' The word 'implemented' was removed from the Config reference.

> Field names below mirror the **implemented** dataclasses in `packages/installer/src/installer/core/model.py` verbatim (snake_case), plus the implemented `Config` dataclass in `packages/installer/src/installer/config.py`. Where a type is not yet built, the row is marked *design-forward* and traces to `installer-design.md`, not to current code.

Author: Here's another place where the docs are _almost_ crossing the line of target state vs. current.  Let's just generalize and remove the "design-forward" tagging throughout this/all the docs.

```suggestion
Field names below mirror the **implemented** dataclasses in `packages/installer/src/installer/core/model.py` verbatim (snake_case), plus the implemented `Config` dataclass in `packages/installer/src/installer/config.py`. This represents target state, not necessarily current code.
```

Author: Applied in commit d387bea — removed all design-forward/implemented tags from ER fields and preamble, replaced with the suggested wording.

```mermaid
erDiagram
    Config      ||--o{ StagingPlan      : "one built per detected Tool"
    StagingPlan ||--o{ StagedItem       : "items: dict keyed by dest_relpath"
    StagedItem  ||--|| Provenance        : "has 1 (origin)"
    StagedItem  ||--o{ FileInclude       : "content flattened from 0..N (transient)"
    StagedItem  ||--o| AllRulesInclude   : "content flattened from ALL-RULES marker (transient; no fields)"
    Config      ||--|| Counters          : "one run tally"
    Config      ||--o{ Orphan            : "prune scan yields 0..N"

    Config {
        Path      home           "destination root (per-tool dest via adapter)"
        ToolTuple tools          "resolved tool set (claude auto-detected; forced via --tools)"
        StrList   plugins        "discovered by scanning src/plugins/"
        bool      dry_run        "--dry-run preview, no writes"
        PruneMode prune          "none | prune | prune_only"
        Path      dump_stage     "--dump-stage target (debug)"
        StrList   retired_globs  "installer.toml [prune].retired"
        Map       tool_overrides "installer.toml [tools]"

>     }

Author: Working as intended — these diagrams describe **target architecture**, not current code. The PR description (updated in the latest push) states this explicitly: *'A component appearing in a diagram without a corresponding file is correct behaviour for an HLD; it is the target the implementation stories are wiring toward.'* Resolving as working-as-intended.

>     }

Author: Working as intended — these diagrams describe **target architecture**, not current code. The PR description (updated in the latest push) states this explicitly: a component or field appearing in a diagram without a corresponding file or implementation is correct for an HLD artifact; it describes where the implementation stories are building toward, not what exists today. Resolving as working-as-intended.

- **`IncludeDirective` is a transient `TypeAlias` union.** `FileInclude | AllRulesInclude`, produced while `templates.py` flattens a `<!-- DYNAMIC-INCLUDE: … -->` marker (file form) or the ALL-RULES marker, then consumed immediately — the flattened text lands in `StagedItem.content`; the directive objects do not survive on the `StagedItem`. `FileInclude` carries the fragment `path`; `AllRulesInclude` carries no fields — it expands the plan's already-staged rules collection, sorted and joined with a `---` separator.
- **`FileKind` is an enum, not an entity.** Its six values are shown inline on `StagedItem.kind`. It is the primary merge-dispatch key; `namespace` is the secondary key (see the dispatch table). Note `Orphan.kind` is a *different*, coarser `Literal["dir", "file"]` — orphan classification only needs dir-vs-file, not the full merge taxonomy — and `Orphan.tool` is a plain `str` (not the `Tool` enum) because the orphan bucket includes plugin namespaces like `beads` that are not tools.

## Merge-dispatch table — `(FileKind, namespace)` → `MergeStrategy`

> The collision matrix — the dispatch contract `core/merge/registry.py` will implement (Epic E; **not yet built**, so this table is the design intent from `installer-design.md`, not a description of current code). It keys on `(FileKind, namespace)`: `NAMESPACED_MD` is the only kind whose namespace changes the strategy; for every other kind the namespace component is unused and the lookup degenerates to a `FileKind`-only key. The strategy names below are the design's; the modules land in Epic E.

Author: ```suggestion
The collision matrix — the dispatch contract `core/merge/registry.py` implements (see `installer-design.md`). It keys on `(FileKind, namespace)`: `NAMESPACED_MD` is the only kind whose namespace changes the strategy; for every other kind the namespace component is unused and the lookup degenerates to a `FileKind`-only key. The strategy names below are the design's; the modules land in Epic E.
```

Author: Applied in commit d387bea — merge-dispatch table intro now uses the suggested wording.

| `Orphan` | A prune candidate: on disk, absent from the plan, matching a retired glob. |
| `Counters` | The per-run tally (`staged` / `created` / `updated` / `skipped` / `pruned` / `backed_up`) surfaced in the exit summary. |
| Canonical ownership | Which actor is the source of truth for a piece of data: the human (source tree), the installer (plan + writes), or the tool (deployed store at runtime). |

## Purpose

Three complementary data views in one file:

1. **The in-memory model** (`Config`, `StagingPlan`, `StagedItem`, `Provenance`, `IncludeDirective`, `Orphan`, `Counters`) as an ER diagram — the shapes the engine builds and passes around. None of these persist; they live and die within one process invocation.
2. **The merge-dispatch table** — the `(FileKind, namespace)` → `MergeStrategy` lookup that the collision matrix is built on.
3. **The config + ownership boundaries** — the `installer.toml` schema, and which actor owns which data as it flows source → memory → disk → runtime.

The data view answers: *what shapes does the installer build in memory, what drives collision resolution, and who owns each piece of data along the way?*

## In-memory model ER diagram

> Field names below mirror the **implemented** dataclasses in `packages/installer/src/installer/core/model.py` verbatim (snake_case), plus the implemented `Config` dataclass in `packages/installer/src/installer/config.py`. This represents target state, not necessarily current code.

Author: Addressed in commit c0cd7f6 — preamble now reads 'Field names mirror the **target** dataclasses… This represents target state, not necessarily current code.' The word 'implemented' was removed from the Config reference.

|---|---|---|---|
| `NAMESPACED_MD` | `rules` | `append_rules` | Join `existing + "\n---\n" + incoming` — rules compose. |
| `NAMESPACED_MD` | `commands` / `skills` / `agents` | `fatal` | **Raise** — two items with the same name is an authoring error; the message names both files. |
| `SETTINGS_JSON` | — | `json_union` | Deep union: nested-dict precedence, array union + sort, type-mismatch surfaced. |
| `JSONC` | — | `last_wins_warn` | Replace, with a warning that an existing file was overwritten. |
| `TOML` | — | `last_wins_warn` | Replace, with a warning. |
| `OTHER` | — | `last_wins_silent` | Replace silently. |
| `DIR` | — | (n/a) | Directories are created, not merged. |

> Formula files are `.toml`, so they classify as `FileKind.TOML` and route via `(TOML, —) → last_wins_warn`. `formulas` is a managed namespace for **backup / prune** routing (it is in the path-aware namespace set), NOT a `NAMESPACED_MD` merge namespace — the only `NAMESPACED_MD` namespaces that change the strategy are `rules` (append) and `commands`/`skills`/`agents` (fatal). The dispatch is data: adding a `(FileKind, namespace)` row is a registry change, not an engine change.

## `installer.toml` schema

> Structured config at `packages/installer/installer.toml`, replacing the legacy `scripts/prune-list` text file at the parity gate. Read once at `Config` build (read-only to the install path).

Author: Working as intended — these diagrams describe **target architecture**, not current code. The PR description (updated in the latest push) states this explicitly: a component or field appearing in a diagram without a corresponding file or implementation is correct for an HLD artifact; it describes where the implementation stories are building toward, not what exists today. Resolving as working-as-intended.
